lib/ansible/galaxy/collection.py

```diff
@@ -1054,6 +1168,54 @@ def _build_dependency_map(collections, existing_collections, b_temp_path, apis,
     return dependency_map
 
 
+def _collections_from_scm(collection, requirement, b_temp_path, force, parent=None):
+    """
+    Returns a list of collections found in the repo. If there is a galaxy.yml in the collection then just return
+    the specific collection. Otherwise, check each top-level directory for a galaxy.yml.
+
+    :param collection: URI to a git repo
+    :param requirement: The version of the artifact
+    :param b_temp_path: The temporary path to the archive of a collection
+    :param force: Whether to overwrite an existing collection or fail
+    :param parent: The name of the parent collection
+    :return: List of CollectionRequirement objects
+    """
+
+    reqs = []
+    name, version, path, fragment = parse_scm(collection, requirement)
+    b_repo_root = to_bytes(name, errors='surrogate_or_strict')
+
+    if fragment:
+        b_fragment = to_bytes(fragment, errors='surrogate_or_strict')
+        b_collection_path = os.path.join(b_temp_path, b_repo_root, b_fragment)
+    else:
+        b_collection_path = os.path.join(b_temp_path, b_repo_root)
+
+    b_galaxy_path = os.path.join(b_collection_path, b'galaxy.yml')
+
+    err = "{0} appears to be an SCM collection source, but the required galaxy.yml was not found. " \
+          "Append #path/to/collection/ to your URI (before the comma separated version, if one is specified) " \
+          "to point to a directory containing the galaxy.yml or directories of collections".format(collection)
+
+    display.vvvvv("Considering {0} as a possible path to a collection's galaxy.yml".format(b_galaxy_path))
+    if os.path.exists(b_galaxy_path):
+        return [CollectionRequirement.from_path(b_collection_path, force, parent, fallback_metadata=True)]
+
+    if not os.path.isdir(b_collection_path) or not os.listdir(b_collection_path):
+        raise AnsibleError(err)
+
+    for b_possible_collection in os.listdir(b_collection_path):
+        b_collection = os.path.join(b_collection_path, b_possible_collection)
+        b_galaxy = os.path.join(b_collection, b'galaxy.yml')
```

Comment:
Currently the code is appending `galaxy.yml` to both files and folders so we should probably skip each entry if it's not a directory.

I just tested this against `ansible-galaxy collection install git+https://github.com/chocolatey/chocolatey-ansible` and it failed due to it not finding the `galaxy.yml` when I expected it to find it at `chocolatey/galaxy.yml`.

Reply:
That example fails for me too, but it seems like it's because the galaxy.yml is not valid for yaml.safe_loads rather than because the `galaxy.yml` can't be found.
```
ERROR! Failed to parse the galaxy.yml at '/home/user/.ansible/tmp/ansible-local-307040vg8x42it/tmpkah0w9xu/chocolatey-ansible/chocolatey/galaxy.yml' with the following error:
while constructing a mapping
  in "b'/home/user/.ansible/tmp/ansible-local-307040vg8x42it/tmpkah0w9xu/chocolatey-ansible/chocolatey/galaxy.yml'", line 12, column 10
found unhashable key
  in "b'/home/user/.ansible/tmp/ansible-local-307040vg8x42it/tmpkah0w9xu/chocolatey-ansible/chocolatey/galaxy.yml'", line 12, column 11
```

Reply:
Yea I came across that after explicitly setting the path but it still fails to find the galaxy.yml without it due to the current logic.

Reply:
Maybe we just need to add a note/warning to the docs saying that the galaxy.yml also needs to be valid for this to load. It might be overkill with all these warnings but I’m trying to cover use cases that people will try and then subsequently raise issues for where hopefully a note will stop some of that.

Reply:
Sounds good, I'll add one

Reply:
@s-hertel this is one of the hardcoded places that I mentioned in the comment below. Probably the last one...

Reply:
@webknjaz Oh, oops. I grepped, but somehow missed that one.